### PR TITLE
Extended indexing operators

### PR DESCRIPTION
**Pull request by Octachron**
After #69, #622 and the last developper meeting, this PR proposes to add a new family of user-definable indexing operators, that does not interfere nor can be confused with any existing indexing operators.

Extended indexing operators names start with a leading dot  `.` followed by an operator character, except `.`(to avoid confusion with `..` which has a different meaning ) or `<` (used as an opening bracket at the type level) , and a possibly empty sequence of operator characters. It must be closed with a couple of enclosing brackets ( i.e. `()`, `[]`, `{}` ) and optionaly an assignment operator `<-`:

```
let (  .%( ) ) p (x,y) = p.(x).(y)
let (  .$?${ } ) p (x,y,z) = p.(x).(y).(z)
let (  .~[ ] <- ) p (x,y,z,t) v = p.(x).(y).(z).(t) <- v 
```

Expressions of the form `expr_1.∘∘∘(expr_2)` are then desugared to ` ( .∘∘∘(  )) expr_1 expr_2 `.
Similarly for assignment, `expr_1.∘∘∘(expr_2) <- expr_3` is desugared to  `( .∘∘∘(  )<-) expr_1 expr_2 expr_3`. The same syntactic sugar is applied to all bracket variants (i.e.  `{`, `(`, `[` ). 

|                   | Expression                                 | Desugaring                                     |
------------- | ---------------------------------|-------------------------------------|
 Access         | expr_1.∘∘∘⟨expr_2⟩                    | ( .∘∘∘⟨  ⟩ ) expr_1 expr_2                 |
 Assignment  | expr_1.∘∘∘⟨expr_2⟩  <- expr_3  | ( .∘∘∘⟨  ⟩ <- ) expr_1 expr_2 expr_3  |

The compulsory operator character should help to mark these operators as user-defined, and decrease the confusion potential of these operators. Moreover, the use of a qualified dot respects the use of  unqualified `.` as a projection primitive.

### CODE REVIEW

**Comment by gasche** [comment]
I like it! `.$?$(...)` is terribly ugly, but `.~[ ]` and `.%( )` both make sense, and I agree it's a good way out of your conflict with type-directed resolution.

(In general I also like the idea of having operator names mirror the operator syntax, similar to Agda's mixfixes, which has first been given to me, in an OCaml contet, by Nicolas Pouillard in 2010. But this was already part of your previous proposal.)

**Comment by gasche** [comment]
If I understand correctly, the testsuite currently checks the parsed AST. Would it be possible to also test that these programs are pretty-printed back correctly (that is, in a way that can be parsed back to the same thing)?

**Review by gasche — APPROVED** [review]

**Comment by Octachron** [comment]
The `parsetree/source.ml` test already checks that the pretty-printed ast can be parsed back.

File: parsing/parser.mly [hunk 1]
```diff
@@ -1407,6 +1408,24 @@ expr:
...
+        mkexp @@ Pexp_apply(id , [Nolabel, $1; Nolabel, $4; Nolabel, $7]) }
+  | simple_expr DOTOP LPAREN expr RPAREN
+      { let id = mkexp @@ Pexp_ident( ghloc @@ Lident ("." ^ $2 ^ "()")) in
+        mkexp @@ Pexp_apply(id, [Nolabel, $1; Nolabel, $4]) }
+  | simple_expr DOTOP LBRACE expr RBRACE LESSMINUS expr
+      { let id = mkexp @@ Pexp_ident( ghloc @@ Lident ("." ^ $2 ^ "{}<-")) in
+        mkexp @@ Pexp_apply(id , [Nolabel, $1; Nolabel, $4; Nolabel, $7]) }
+  | simple_expr DOTOP LBRACE expr RBRACE
+      { let id = mkexp @@ Pexp_ident( ghloc @@ Lident ("." ^ $2 ^ "{}")) in
+        mkexp @@ Pexp_apply(id, [Nolabel, $1; Nolabel, $4]) }
```
Comment (Drup) [inline parsing/parser.mly:1436]: This is orthogonal, but it might be beneficial to add rules of the form `simple_expr DOTOP LBRACE expr error` to give better error messages.

**Comment by Drup** [comment]
The patch looks good to me (and I like the proposed solution).

I wonder if maybe it would be time to change the representation of (non module) identifier to a sum type. That would make the printing implementation slightly less ridiculous.

File: parsing/parser.mly (hunk 1 above)
Comment (Octachron) [reply parsing/parser.mly:1436]: Good point, fixed.

**Comment by Octachron** [comment]
Extending on @Drup comment, I have added a comment to clarify the cryptic string manipulation in `parsing/pprintast.ml` which extract operator informations from its name.

**Comment by damiendoligez** [comment]
Looks good but I don't much like the restrictions on characters:

* for `<` : is there a real reason to disallow it (i.e. a conflict in the grammar) ?
* for `.` : of course `..` is a keyword, but I don't think that's a reason to disallow all "operators" that begin with `..`.

**Comment by Octachron** [comment]
With the current scheme (i.e. using a new DOTOP token in the lexer) allowing either `.` or `<` as starting characters would require modification in unrelated part of the parser due to the use of `<` and `>` as brackets for object type:
* For `<`,  in `let f: 'a .< x:'a> -> 'a= … `, `.<` will be read as a `DOTOP(<)` token.
* For `.`, in `let f: <x:'a; ..> -> 'a = …`,  `..>` will be read as a `DOTOP(.>)` token.

There is potentially two ways to get around this limitation: restrict the syntax to merge `.%(` and friends in a single token rather than two (`.%` `(`) (a solution that was suggested to me by @Drup earlier) or generalize the syntax by removing the lexer modification and implementing it as a grammar modification only.

The first solution will slightly break the symmetry between classical indexing operators and the extended ones: `a   .   (b)` will be valid, but not `a    .%    (b)`. The last solution will allow more whitespace at strange positions: `a    ..  . ..   %   (b) `.

**Comment by damiendoligez** [comment]
OK, those are good reasons.

**Comment by Octachron** [comment]
What is the status of this PR for 4.0{6,7}?

If the restriction on the operators character is deemed too complicated, a coarser and simpler restriction may be to forbid the characters `<>` with the motivation that they are too close typographically from the left and right angle brackets `⟨⟩` and thus confusing in for instance `x.<?>.(y)`.

**Comment by damiendoligez** [comment]
I think the restriction is OK. It will always be possible to write unreadable programs so we shouldn't spend too much time on this "issue".

Please rebase and I'll merge.

**Comment by lpw25** [comment]
I just noticed that this doesn't support things like `x.Foo.%(...)` -- analogous to `x.Foo.label`. This is a shame. I guess its too late to get that in before the freeze. Maybe Damien will generously consider fixing this omission a bug fix?

**Comment by mshinwell** [comment]
I took the liberty of rebasing this myself.  Will be ready to merge once Travis etc finishes.

**Comment by damiendoligez** [comment]
>Maybe Damien will generously consider fixing this omission a bug fix?

Depending on the complexity of the fix and the proximity of the release, maybe.

**Comment by Octachron** [comment]
@lpw25 , I will send a PR adding back qualified operators tomorrow morning.

File: Changes [hunk 2]
```diff
@@ -12,10 +18,11 @@ Working version
...
 
-- GPR#792: fix limitations of destructive substitutions, by
-  allowing "S with type t := type-expr",
-  "S with type M.t := type-expr", "S with module M.N := path"
-  (Valentin Gatien-Baron, review by Jacques Garrigue and Leo White)
+- GPR#1064: extended indexing operators, add a new class of
+  user-defined indexing operators, obtained by adding at least
+  one operator character after the dot symbol to the standard indexing
+  operators: e,g ".%()", ".?[]", ".@{}<-"
+  (Florian Angeletti, review by Damien Doligez and Gabriel Radanne)
```
Comment (nojb) [inline Changes:25]: @Octachron I think this PR breaks backwards compatibility in some edge cases: see https://caml.inria.fr/mantis/view.php?id=7637.

**Comment by yallop** [comment]
Not necessarily a problem, but worth noting here: this change uses up the syntactic space that was [considered](https://caml.inria.fr/mantis/view.php?id=5980#c9278) for the succinct version of overriding local open, i.e.

```ocaml
M.!(e)
```

as a shorthand for 

```ocaml
let open! M in e
```

With this PR, `M.!(e)` is instead interpreted as `(.!()) M e`.

**Comment by Drup** [comment]
@yallop You mean, the one that was rejected (https://github.com/ocaml/ocaml/pull/218) ? :)

**Comment by yallop** [comment]
Another issue with this change (which I should have spotted earlier): it reserves `.~`, which has been [used by MetaOCaml](http://okmij.org/ftp/ML/MetaOCaml.html#using) for splicing for a long time.

However, it appears that none of the code available via OPAM is yet using `.~`, so it's not too late to restore compatibility.  I'll open a new pull request with more details (update: https://github.com/ocaml/ocaml/pull/2106).
